template/steps/finetune.py

```diff
@@ -1,234 +1,126 @@
+            tokenizer, mlm=False
+        ),
+        callbacks=[ZenMLCallback()],
     )
-    train_args = TrainArgs(**config.training.dict())
-    eval_args = EvalArgs(**config.eval.dict())
-    setup(
-        devices=1,
-        io=io_args,
-        train=train_args,
-        eval=eval_args,
-        precision=config.precision,
-        quantize=config.quantize,
-        **config.lora.dict(),
+
+    model.config.use_cache = (
+        False  # silence the warnings. Please re-enable for inference!
     )
 
-    if config.merged_output_repo:
-        lora_path = adapter_output_dir / "lit_model_lora_finetuned.pth"
-
-        merge_output_dir = (
-            Path("output/lora_merged") / dataset_name / model_name
-        )
-        merge_lora(
-            lora_path=lora_path,
-            checkpoint_dir=checkpoint_dir,
-            out_dir=merge_output_dir,
-            precision=config.precision,
-            **config.lora.dict(),
-        )
-
-        for path in Path(checkpoint_dir).glob("*.json"):
-            destination = Path(merge_output_dir) / path.name
-            shutil.copy(src=path, dst=destination)
-
-        if config.convert_to_hf_checkpoint:
-            upload_dir = (
-                Path("output/lora_merged_hf") / dataset_name / model_name
-            )
-            upload_dir.mkdir(parents=True, exist_ok=True)
-            convert_lit_checkpoint(
-                checkpoint_path=config.merged_output_repo / "lit_model.pth",
-                config_path=config.merged_output_repo / "lit_config.json",
-                output_path=upload_dir / "pytorch_model",
-            )
-        else:
-            upload_dir = merge_output_dir
-
-        commit = upload_folder(
-            repo_id=config.merged_output_repo,
-            folder_path=upload_dir,
-            token=access_token,
-        )
-        log_model_metadata(
-            metadata={
-                "merged_model_huggingface_commit_hash": commit.oid,
-                "merged_model_huggingface_commit_url": commit.commit_url,
-            }
-        )
-
-    if config.adapter_output_repo:
-        commit = upload_folder(
-            repo_id=config.adapter_output_repo,
-            folder_path=adapter_output_dir,
-            token=access_token,
-        )
-        log_model_metadata(
-            metadata={
-                "adapter_huggingface_commit_hash": commit.oid,
-                "adapter_huggingface_commit_url": commit.commit_url,
-            }
-        )
-        return None
-    else:
-        # If the adapter should not be uploaded to the HF Hub, we store it
-        # in the artifact store
-        return adapter_output_dir
+    logger.info("Training model...")
+    trainer.train()
+
+    logger.info("Saving model...")
+    model.config.use_cache = True
+    ft_model_dir = Path("model_dir")
+    ft_model_dir.mkdir(parents=True, exist_ok=True)
+    trainer.save_model(ft_model_dir)
+
+    return ft_model_dir
```

Comment:
So this pattern also somehow bypasses the MCP, then, right? If this is something we're encouraging people to do, we might want to consider (for future) having the MCP know and understand paths to models?

Reply:
Yeah, this seems to be on the intersection of Mounts + better HuggingFace/Accelerate integration in the core. I don't see this as a relevant thing to do here.